File: pkg/core/security/security.go
// Package security
package security

import (
	"crypto"
	"crypto/rand"
)

type Certificate [2048]byte

func GenerateCertificate() (Certificate, error) {
	var cert Certificate
	_, err := rand.Read(cert[:])
	if err != nil {
		return Certificate{}, err
	}
	return cert, nil
}

func (cert *Certificate) ID() string {
	return string(crypto.SHA256.New().Sum(cert[:]))
}

type PairToken [128]byte

func GeneratePairToken() (PairToken, error) {
	var token PairToken
	_, err := rand.Read(token[:])
	if err != nil {
		return PairToken{}, err
	}
	return token, nil
}

File: nlem/src/security.rs
pub const CERTIFICATE_SIZE: usize = 8; // has no use for now
pub type Certificate = Vec<u8>;
pub type DeviceID = [u8; 16];
pub type PairToken = [u8; 32];

pub fn array_from_vec<const N: usize>(mut v: Vec<u8>) -> [u8; N] {
    v.resize(N, 0u8);
    v.try_into().unwrap()
}

pub fn generate_id() -> DeviceID {
    let mut devid: DeviceID = [0u8; _];
    rand::fill(&mut devid);
    devid
}
pub fn generate_certificate() -> Certificate {
    let mut cert = vec![0u8; CERTIFICATE_SIZE];
    rand::fill(&mut cert);
    cert
}

pub fn generate_pair_token() -> PairToken {
    let mut token: PairToken = [0u8; _];
    rand::fill(&mut token);
    token
}
